src/unity/python/turicreate/toolkits/drawing_classifier/drawing_classifier.py

```diff
@@ -194,39 +206,72 @@ def update_progress(cur_loss, iteration):
     model.hybridize()
     trainer = _mx.gluon.Trainer(model.collect_params(), 'adam')
 
-    train_loss = 0.
-    for batch in loader:
-        data = _mx.gluon.utils.split_and_load(batch.data[0], 
-            ctx_list=ctx, batch_axis=0)[0]
-        label = _mx.nd.array(
-            _mx.gluon.utils.split_and_load(batch.label[0], 
-                ctx_list=ctx, batch_axis=0)[0]
-            )
-
+    train_accuracy = _mx.metric.Accuracy()
+    validation_accuracy = _mx.metric.Accuracy()
+    
+    def compute_validation_accuracy(validation_accuracy):
+        validation_loader.reset()
+        for validation_batch in validation_loader:
+            validation_batch_data = _mx.gluon.utils.split_and_load(
+                validation_batch.data[0], ctx_list=ctx, batch_axis=0)
+            validation_batch_label = _mx.gluon.utils.split_and_load(
+                validation_batch.label[0], ctx_list=ctx, batch_axis=0)
+            validation_outputs = []
+            for x, y in zip(validation_batch_data, validation_batch_label):
+                z = model(x)
+                validation_outputs.append(z)
+            validation_accuracy.update(
+                validation_batch_label, validation_outputs)
+
+    for train_batch in train_loader:
+        train_batch_data = _mx.gluon.utils.split_and_load(
+            train_batch.data[0], ctx_list=ctx, batch_axis=0)
+        train_batch_label = _mx.gluon.utils.split_and_load(
+                train_batch.label[0], ctx_list=ctx, batch_axis=0)
+        train_outputs = []
+
+        # Inside training scope
         with _autograd.record():
-            output = model(data)
-            loss = softmax_cross_entropy(output, label)
-        loss.backward()
-        # update parameters
-        trainer.step(1)
+            for x, y in zip(train_batch_data, train_batch_label):
+                z = model(x)
+                # Computes softmax cross entropy loss.
+                loss = softmax_cross_entropy(z, y)
+                # Backpropagate the error for one iteration.
+                loss.backward()
```

Comment:
I'm implementing similar functionality for the sound classifier and in my testing I noticed an interesting bug. I believe your code also has the same issue. Perhaps it's also in other toolkits as well....

Try passing in the same SFrame to `create(...)` as both the `input_dataset` and the `validation_set`. The training accuracy and the validation accuracy columns that get printed out should match, since they're the same SFrame, but I don't think they will.

I don't think it's correct to save the predicted labels while you're still doing backward passes. You probably want to do all your backward passes for in one loop and then iterate over your training set again and save the predictions.

Reply:
Fixed. Thanks for pointing this out!